src/osd/ClassHandler.cc

```diff
@@ -312,21 +304,24 @@ void ClassHandler::ClassFilter::unregister()
 
 int ClassHandler::ClassMethod::exec(cls_method_context_t ctx, bufferlist& indata, bufferlist& outdata)
 {
-  int ret;
-  if (cxx_func) {
-    // C++ call version
-    ret = cxx_func(ctx, &indata, &outdata);
-  } else {
-    // C version
-    char *out = NULL;
-    int olen = 0;
-    ret = func(ctx, indata.c_str(), indata.length(), &out, &olen);
-    if (out) {
-      // assume *out was allocated via cls_alloc (which calls malloc!)
-      buffer::ptr bp = buffer::claim_malloc(olen, out);
-      outdata.push_back(bp);
+  int ret = 0;
+  std::visit([&](auto method) {
+    using method_t = decltype(method);
+    if constexpr (std::is_same_v<method_t, cls_method_cxx_call_t>) {
+      // C++ call version
+      ret = method(ctx, &indata, &outdata);
+    } else if constexpr (std::is_same_v<method_t, cls_method_call_t>) {
+      // C version
+      char *out = nullptr;
+      int olen = 0;
+      ret = method(ctx, indata.c_str(), indata.length(), &out, &olen);
+      if (out) {
+        // assume *out was allocated via cls_alloc (which calls malloc!)
+        buffer::ptr bp = buffer::claim_malloc(olen, out);
+        outdata.push_back(bp);
+      }
```

Comment:
This a very compact visitor! :-) The last minor I can think about is not ignoring unrecognized type in the `std::variant`. The snippet bases on idea mentioned on [cppreference.com](https://en.cppreference.com/w/cpp/language/if).

```cpp
template<class T> struct dependent_false : std::false_type {
};
int ClassHandler::ClassMethod::exec(cls_method_context_t ctx, bufferlist& indata, bufferlist& outdata)
{
  // ...
  std::visit([&](auto method) {
    using method_t = decltype(method);
    if constexpr (std::is_same_v<method_t, cls_method_cxx_call_t>) {
      // ...
    } else {
      static_assert(dependent_false<T>::value, "Type must be recognized");
    }
  }
```

Reply:
@rzarzynski updated and repushed. i am using a different implementation `dependent_false` for less LOC.